src/Void.Core/BaseMessageTypes.fs

```diff
@@ -5,36 +5,43 @@ type Message = interface end
     let noMessage = NoMessage :> Message
 
-type NoResponseToRequest<'TRequest when 'TRequest :> RequestMessage> =
-    {
-        Request : 'TRequest
-    }
-    interface Message
-
 type Handle<'TMsg when 'TMsg :> Message> =
     'TMsg -> Message
 
-type HandleRequest<'TRequest when 'TRequest :> RequestMessage> = 
-    'TRequest -> ResponseMessage<'TRequest>
+type HandleRequest<'TRequest, 'TResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest>> = 
+    'TRequest -> 'TResponse
+
+type MaybeHandleRequest<'TRequest, 'TResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest>> = 
+    'TRequest -> 'TResponse option
+
+type HandlePackagedRequest<'TRequest, 'TResponse, 'TPackagedRequest, 'TPackagedResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest> and 'TPackagedRequest :> EnvelopeMessage<'TRequest> and 'TPackagedResponse :> EnvelopeMessage<'TResponse>> = 
+    'TPackagedRequest -> 'TPackagedResponse
+
+type MaybeHandlePackagedRequest<'TRequest, 'TResponse, 'TPackagedRequest, 'TPackagedResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest> and 'TPackagedRequest :> EnvelopeMessage<'TRequest> and 'TPackagedResponse :> EnvelopeMessage<'TResponse>> = 
+    'TPackagedRequest -> 'TPackagedResponse option
 
-type MaybeHandleRequest<'TRequest when 'TRequest :> RequestMessage> = 
-    'TRequest -> ResponseMessage<'TRequest> option
+type RequestSender =
+    abstract member makeRequest<'TRequest, 'TResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest>> : 'TRequest -> 'TResponse option
 
-type HandleResponse<'TRequest, 'TResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest>> = 
-    'TResponse -> Message
+type PackagedRequestSender =
+    abstract member makePackagedRequest<'TRequest, 'TResponse, 'TPackagedRequest, 'TPackagedResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest> and 'TPackagedRequest :> EnvelopeMessage<'TRequest> and 'TPackagedResponse :> EnvelopeMessage<'TResponse>> : 'TPackagedRequest -> 'TPackagedResponse option
 
-type SubscribeToBus =
+type Bus =
+    abstract member publish : Message -> unit
     abstract member subscribe<'TMsg when 'TMsg :> Message> : Handle<'TMsg> -> unit
-    abstract member subscribeToRequest<'TRequest when 'TRequest :> RequestMessage> : HandleRequest<'TRequest> -> unit
-    abstract member subscribeToRequest<'TRequest when 'TRequest :> RequestMessage> : MaybeHandleRequest<'TRequest> -> unit
-    abstract member subscribeToResponse<'TRequest, 'TResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest>> : HandleResponse<'TRequest, 'TResponse> -> unit
\ No newline at end of file
+    abstract member subscribeToRequest<'TRequest, 'TResponse when 'TRequest :> RequestMessage and 'TResponse :> ResponseMessage<'TRequest>> : HandleRequest<'TRequest, 'TResponse> -> unit

```

Comment:
I don't think so. Is there a disadvantage to overloading that you are concerned about?